Fix generation of RazorOutputFiles ItemGroup with duplicated views.

gamulka opened the pull request:
While testing the newest NuGet release (2.4.2), I've found my builds were failing with (tons of) errors like:

`Skunk.WebSite/obj/CodeGen/Features/Account/AccountEnroll.cshtml.vb' could not be found

After inspecting msbuild's output in detail, I've found that every view in my proyect was being passed as twice to csc's command line, first with .cshtml.cs extension, and them with .cshtml.vb.

It looks like recent VB support (#85) introduced a regression into RazorGenerator.MSBuild.targets, by which RazorOutputFiles ItemGroup was invoked twice, frist for CS files, then for VB, thus, overwritting the initial (CS) declaration.

This pull-req changes RazorGenerator.MSBuild.targets, so it accounts for each kind of view (CS vs VB) separatelly, and emitting RazorSrcFiles && RazorOutputFiles as an union/concat of filepaths of each kind.

davidebbo commented:
Ok, try 2.4.3! I'll keep 2.4.2 unlisted.

gamulka commented:
@davidebbo / @pranavkm this new version 2.4.2 is working properly :)

dove commented:
hi @gamulka @davidebbo I just updated to 2.4.3 from 2.4.1 and my C# project does not build successfully.  Error states there are illegal characters in path of target file.  Would this be related to changes here?

I've reverted to 2.4.1 for now.  Happy to provide more info if you want it.  Also can look at fix but need to brief myself more on the workings of this.

gamulka commented:
Hi,

My you provide a link to a zip file containing a failing sample proyect?
El 5/5/2016 12:16, "Colum Clissmann" <notifications@github.com> escribió:

> hi @gamulka <https://github.com/gamulka> @davidebbo
> <https://github.com/davidebbo> I just updated to 2.4.3 from 2.4.1 and my
> C# project does not build successfully. Error states there are illegal
> characters in path of target file. Would this be related to changes here?
>
> I've reverted to 2.4.1 for now. Happy to provide more info if you want it.
> Also can look at fix but need to brief myself more on the workings of this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/RazorGenerator/RazorGenerator/pull/90#issuecomment-217118549>
>

dove commented:
I'd assume it has something to do with my project setup otherwise the test projects would fail.  

I'm looking at it now, installed 2.4.3 again and still get illegal chars error.  However, if I remove all vb file checks from targets file it builds fine, see below for modified file.  

I thought I maybe had a vb file somewhere in project that caused confusion but that's not the case.   I don't have vb installed if that makes a difference?

I'll keep investigating but thought you might have an idea in meantime?

<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <PropertyGroup>
        <PrecompileRazorFiles Condition=" '$(PrecompileRazorFiles)' == '' ">true</PrecompileRazorFiles>
        <RazorGeneratorMsBuildPath Condition=" '$(RazorGeneratorMsBuildPath)' == '' ">$(MSBuildThisFileDirectory)\..\tools\RazorGenerator.MsBuild.dll</RazorGeneratorMsBuildPath>
        <RazorViewsCodeGenDirectory Condition=" '$(RazorViewsCodeGenDirectory)' == '' ">$(MsBuildProjectDirectory)\obj\CodeGen\</RazorViewsCodeGenDirectory>
      
        <CompileDependsOn Condition=" '$(PrecompileRazorFiles)' == 'true' ">
            PrecompileRazorFiles;
            $(CompileDependsOn);
        </CompileDependsOn>
    </PropertyGroup>

    <Target Name="_ResolveRazorFiles">
        <ItemGroup>
            <RazorCsSrcFiles Condition=" '@(RazorCsSrcFiles)' == '' and '%(Extension)' == '.cshtml' "
                Include="@(Content);@(None)" />
            <RazorCsSrcFiles Condition=" '@(RazorCsSrcFiles)' == '' "
                Include="**\*.cshtml" />
            <RazorOutputFiles
                Include="@(RazorCsSrcFiles -> '$(RazorViewsCodeGenDirectory)%(RelativeDir)%(Filename)%(Extension).cs')" />
        </ItemGroup>
  
         <ItemGroup>
             <RazorSrcFiles Include="@(RazorCsSrcFiles)" />
             <RazorOutputFiles Include="@(RazorCsOutputFiles)" />
         </ItemGroup>
    </Target>

    <UsingTask AssemblyFile="$(RazorGeneratorMsBuildPath)" TaskName="RazorCodeGen" />
    <Target Name="PrecompileRazorFiles"
                    DependsOnTargets="_ResolveRazorFiles"
                    Inputs="@(RazorSrcFiles)"
                    Outputs="@(RazorOutputFiles)">
        <RazorCodeGen ProjectRoot="$(MsBuildProjectDirectory)"
                                    FilesToPrecompile="@(RazorSrcFiles)"
                                    CodeGenDirectory="$(RazorViewsCodeGenDirectory)"
                                    RootNamespace="$(RootNamespace)">
            <Output TaskParameter="GeneratedFiles" ItemName="FilesGenerated" />
        </RazorCodeGen>
        <ItemGroup>
          <Compile Include="@(FilesGenerated)" Condition=" '@(RazorOutputFiles)' == '' "/>
          <Compile Include="@(RazorOutputFiles)" />
        </ItemGroup>
    </Target>
</Project>

davidebbo commented:
I have unlisted `2.4.3` until this is figured out.

I probably won't have time to look into this myself, so I'd appreciate if someone is able to get to the bottom of it. Thanks!

pruiz commented:
@dove What's the exact error your are getting from msbuild while building?

pruiz commented:
@dove Also, a sample project, which we can use in order to reproduce the issue, would help a lot.

dove commented:
If I create a clean project with just the generator it works fine.  The project in question is large enough and is not open for sharing.  The error I see is "**Illegal characters in path**."  I tried running with TraceDesignTime set to true but didn't get any more detail in the compile logs.

Trying to isolate the issue, if I remove the RazorVbSrcFiles, i.e.
```
            <RazorSrcFiles Include="@(RazorCsSrcFiles)" />
             <RazorOutputFiles Include="@(RazorCsOutputFiles);@(RazorVbOutputFiles)" />
```
I get the following errors

File name MyProjectDetailsRemoved..obj\CodeGen\**\*.vbhtml.vb' is empty, contains invalid characters, has a drive specification without an absolute path, or is too long
Source File obj\CodeGen\**\*.vbhtml.vb' could not be found

I'll take a look for anything in my project file that might be clashing.  It'd probably be quite a job getting a sample project to replicate.  If I do do you think it's more my project settings or other packages that is worth considering first?

gamulka commented:
@dove I've tried to reproduce it a few times (like having paths with spaces, or paths with some local chars), but in the end, I've been unable to.

I am afraid, without a sample test demonstrating the issue, reproducing this is going to take too much time/resources.

May you try to reduce, remove code/files from your failing project, somehow, so you can provide us a repro?

Otherwise I think we won't be able to help you

davidebbo commented:
Given that it doesn't seem common, I've re-listed `v2.4.3` in nuget. If someone else runs into this, hopefully we'll get a repro and get to the bottom of it.

dove commented:
Fair enough, I've just commented out vb variables for now and am running 2.4.3. will get more details or repo when I have a chance.

wizzardmr42 commented:
@davidebbo I can't see 2.4.3 listed?

davidebbo commented:
@wizzardmr42 that's strange, it's there in the list: http://www.nuget.org/packages/RazorGenerator.MsBuild/. Only 2.4.2 is unlisted.

wizzardmr42 commented:
@davidebbo  sorry - was looking at RazorGenerator.Mvc !

dove commented:
@gamulka updated to 2.4.4 and given commits it's not surprising I still have my issue.  to give a little more info I get a build error from the target file on line that is

<RazorCodeGen ProjectRoot="$(MsBuildProjectDirectory)"
                                    FilesToPrecompile="@(RazorSrcFiles)"
                                    CodeGenDirectory="$(RazorViewsCodeGenDirectory)"
                                    RootNamespace="$(RootNamespace)">

I added a target dependency to output the variables and have the following which are the same when it works as well. 
 _DisplayMessages
 MsBuildProjectDirectory = C:\Source\DevDig\Epic\trunk\app\Epic.Web'
 RazorSrcFiles = ''
 RazorViewsCodeGenDirectory = 'C:\Source\DevDig\Epic\trunk\app\Epic.Web\obj\CodeGen\'
 RootNamespace = 'Epic.Mvc'
 RazorCsSrcFiles = ''
 RazorVbSrcFiles = ''
 RazorOutputFiles = ''
 RazorVbOutputFiles = ''
 RazorCsOutputFiles = ''
 Extension = ''

It's clearly something to do with my setup so will try and isolate further,. just thought I'd post the update, I understand there's not a lot you can do with this.

dove commented:
Looking into the verbose logs I see 

 FilesToPrecompile
```
Areas\aaa\Views\a\Long.cshtml
Areas\bb\Views\a\List.cshtml
Areas\etc\Views\a\CSharp.cshtml
 **\*.vbhtml 
```

That last file looks wrong.  I added a dummy vbhtml view to my project and then it builds past this part as it has an actual file listed.  The project still fails to build but because of errors compiling the vb code.

Can anyone verify this works on Visual Studio 2015 without vb files and without VB language installed for visual studio?

davidebbo commented:
Maybe "_without VB language installed for visual studio_" is the key. Most people don't use VB, but still have it installed since it's in the default setup.

But what's interesting is that you wrote above "_If I create a clean project with just the generator it works fine_". Why wouldn't it have the same issue?

dove commented:
Good point, I installed VB for visual studio and still have same issue.

If you turn on verbose logging for build do you see `**\*.vbhtml `in the list of `FileToPrecompile`, that would give some meaning to the _Illegal characters in path_ error I get, as in it's not a valid file but a matching.  I'm thinking that should not be appended to list of files at all.

davidebbo commented:
So you see this in the verbose log for the project that's broken. Did you try whether you see it in your clean test project that's not broken?

josiahdecker commented:
I'm having the same issue. I think this may be a regression of issue #3. When I remove
`<RazorVbSrcFiles Condition=" '@(RazorVbSrcFiles)' == '' "
                Include="**\*.vbhtml" />`
From the targets file the error goes away.  I believe this occurs because there are no vbhtml files included in the project and there are path lengths in the project folder longer than 255 chars (due to files in the node_modules folder).

dove commented:
That would match my problem, no vbhtml files, node_modules there and the illegal chars error would be thrown by an illegal path (>255 chars).

davidebbo commented:
@josiahdecker it would be great if you could share a repository that repros this issue. So far, the only repro has been from @dove and his cannot be shared.

davidebbo commented:
Taking conversation to #99, since we shouldn't be discussing this issue in a PR. :) We now have 3 reports of this, but still no repro. :(
